File: dox2dox_app/lib/data/models/conversation.dart
import '../../core/enums/app_enums.dart';

class Conversation {
  final String id;
  final String orgId;
  final ConversationType type;
  final String? name;
  final String? createdBy;
  final int? disappearAfterSec;
  final DateTime createdAt;
  final DateTime updatedAt;
  final List<String> memberIds;
  final DateTime? lastMessageAt;
  final String? lastMessagePreview;
  final String? lastMessageSenderId;
  final MessageType? lastMessageType;
  final int unreadCount;

  const Conversation({
    required this.id,
    required this.orgId,
    required this.type,
    required this.name,
    required this.createdBy,
    required this.disappearAfterSec,
    required this.createdAt,
    required this.updatedAt,
    required this.memberIds,
    required this.lastMessageAt,
    required this.lastMessagePreview,
    required this.lastMessageSenderId,
    required this.lastMessageType,
    required this.unreadCount,
  });

  factory Conversation.fromJson(Map<String, dynamic> j) => Conversation(
        id: j['id'] as String,
        orgId: j['org_id'] as String,
        type: ConversationType.fromWire(j['type'] as String),
        name: j['name'] as String?,
        createdBy: j['created_by'] as String?,
        disappearAfterSec: j['disappear_after_sec'] as int?,
        createdAt: DateTime.parse(j['created_at'] as String),
        updatedAt: DateTime.parse(j['updated_at'] as String),
        memberIds: ((j['member_ids'] ?? []) as List).map((e) => e.toString()).toList(),
        lastMessageAt: j['last_message_at'] != null
            ? DateTime.parse(j['last_message_at'] as String)
            : null,
        lastMessagePreview: j['last_message_preview'] as String?,
        lastMessageSenderId: j['last_message_sender_id'] as String?,
        lastMessageType: j['last_message_type'] != null
            ? MessageType.fromWire(j['last_message_type'] as String)
            : null,
        unreadCount: (j['unread_count'] ?? 0) as int,
      );
}

File: doqto_backend/app/schemas/conversation.py
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from app.core.enums import ConversationAccess, ConversationType, MessageType
from app.schemas.common import ORMModel


class ConversationCreateIn(BaseModel):
    type: ConversationType
    name: str | None = Field(default=None, max_length=100)
    member_ids: list[uuid.UUID] = Field(min_length=1)


class ConversationSettingsIn(BaseModel):
    disappear_after_sec: int | None = None


class ConversationAddMembersIn(BaseModel):
    user_ids: list[uuid.UUID] = Field(min_length=1)


class ConversationOut(ORMModel):
    id: uuid.UUID
    org_id: uuid.UUID | None  # NULL = network (cross-org) conversation
    type: ConversationType
    name: str | None
    created_by: uuid.UUID | None
    # Networking tier + provenance (M3, additive). is_network mirrors org_id IS
    # NULL (populated from the model's is_network property).
    access: ConversationAccess = ConversationAccess.OPEN
    initiator_id: uuid.UUID | None = None
    is_network: bool = False
    # Requests filter only (M4, compute-on-read): a low-quality request the
    # client hides from the badge / surfaces separately. Always False elsewhere.
    is_hidden: bool = False
    disappear_after_sec: int | None
    created_at: datetime
    updated_at: datetime
    member_ids: list[uuid.UUID] = []
    display_name: str | None = None  # direct chats: the other member's full name
    last_message_at: datetime | None = None
    last_message_preview: str | None = None
    last_message_sender_id: uuid.UUID | None = None
    last_message_type: MessageType | None = None
    unread_count: int = 0
